src/goto-instrument/reachability_slicer.cpp

```diff
@@ -71,11 +109,71 @@ void reachability_slicert::fixedpoint_from_assertions(
   const is_threadedt &is_threaded,
   slicing_criteriont &criterion)
 {
-  std::vector<cfgt::node_indext> src = get_sources(is_threaded, criterion);
+  // Stack entries are pairs of node indices and a boolean flag that indicates
+  // whether the function's callsite is known, in which case it has already
+  // been placed on the stack and return sites are a no-op.
+
+  std::vector<std::pair<cfgt::node_indext, bool>> stack;
+  std::vector<cfgt::node_indext> sources = get_sources(is_threaded, criterion);
+  for(const auto source : sources)
+    stack.emplace_back(source, false);
+
+  while(!stack.empty())
+  {
+    auto index = stack.back().first;
+    auto callsite_is_known = stack.back().second;
+    stack.pop_back();
 
-  const std::vector<cfgt::node_indext> reachable = cfg.get_reachable(src, true);
-  for(const auto index : reachable)
-    cfg[index].reachable_from_assertion = true;
+    auto &node = cfg[index];
+    if(node.reachable_from_assertion)
+      continue;
+    node.reachable_from_assertion = true;
+
+    if(node.PC->is_function_call())
+    {
+      // Queue the instruction's natural successor (function head, or next
+      // instruction if the function is bodyless)
+      INVARIANT(node.out.size() == 1, "Call sites should have one successor");
```

Comment:
Is this true also for GOTO obtained from a `C` program (e.g. a call via a pointer)?

Reply:
`cfgt` requires that `remove_function_pointer` has run (it ignores function calls that don't call an `ID_symbol`)